### ISSUE TITLE
Exceptions in Fortran

### ISSUE THREAD

**Issue by aradi** [j3-fortran/fortran_proposals#236]
Based on the [discussion on fortran-lang.discourse](https://fortran-lang.discourse.group/t/errorfx-fortran-exception-library/2072) about the [ErrorFx](https://github.com/aradi/errorfx) library, a mechanism for exception like error handling should be created.

ErrorFx demonstrates, that such a mechanism is possible with current Fortran already. But it needs a lot of boiler plate code, which in ErrorFx are currently substituted by Fypp-macros. I drafted a possible syntax for all the scenarios we can already handle in ErrorFx. I'd also provide equivalent Fortran code, but probably we can already start discussion, in case somebody sees some general flaws in the concept.

```fortran
! All errors are extensions of a base type
type, extends(fatal_error) :: io_error
  character(:), allocatable :: name
  integer :: unit = - 1
end type io_error

type, extends(fatal_error) :: allocation_error
  integer :: allocation_size
end type allocation_error
  

subroutine subroutine_with_error(...) throws(io_error, allocation_error)
  ...
  throw io_error(name="somefile.dat")
 ...
end subroutine subroutine_with_error


function function_with_error(...) result(...) throws(io_error)
  ...
  throw io_error(unit=9)
  ...
end function function_with error


! Propagating error upwards
try call subroutine_with_error()

! Propagating error upwards in function call
i = try function_with_error()

! Catching if any error thrown and assign a default value in that case
i = try function_with_error() else -1

! The classical try-catch block
! errorvar is the name of the local variable representing the error which was caught
! If an error is not handled here, it will be automatically propagated upwards
try catch (errorvar)
  i = function_with_error
  ! or altenatively
  call subroutine_with_error()
catch (io_error)
  ! do something
  print *, "UNIT:", errorvar%unit
  ! We pass it upwards as a more generic error
  throw fatal_error :: errorvar
catch (some_other_error)
  ! do something else for this error
end try catch

! A call to a subroutine or a function with error outside of a try should trigger a compiler error
call subroutine_with_error()  ! should trigger compiler error
i = function_with_error()   ! should trigger compiler error as well
```

**Comment by certik** [comment]
@aradi thanks for opening up an issue for this to start a discussion.

So the single line `try` syntax is very good:
```
! Propagating error upwards
try call subroutine_with_error()

! Propagating error upwards in function call
i = try function_with_error()

! Catching if any error thrown and assign a default value in that case
i = try function_with_error() else -1
```
It greatly simplifies error handling, but it is just nicer syntax for what can be done already, as you have shown with your ErrorFx library. For reference, to make this proposal standalone, the proposal is that this code:
```fortran
subroutine routine_propagating_error(..., error)
  ...
  type(fatal_error), allocatable, intent(out) :: error
  ...
  ! If error happend, we propagate it upwards, otherwise we continue
  try call routine_with_possible_error(...)
  print "(a)", "Apparently no error occured"
  ...
end subroutine routine_propagating_error
```
is equivalent to:
```fortran
subroutine routine_propagating_error(..., error)
  ...
  type(fatal_error), allocatable, intent(out) :: error
  ...
  call routine_with_possible_error(..., error)
  ! If error happend, we propagate it upwards, otherwise we continue
  if (allocated(error)) return
  print "(a)", "Apparently no error occured"
  ...
end subroutine routine_propagating_error
```
There is no "magic" behind the scene.

Can you write what this code is equivalent to:
```fortran
try catch (errorvar)
  i = function_with_error
  ! or altenatively
  call subroutine_with_error()
catch (io_error)
  ! do something
  print *, "UNIT:", errorvar%unit
  ! We pass it upwards as a more generic error
  throw fatal_error :: errorvar
catch (some_other_error)
  ! do something else for this error
end try catch
```
I think we have not discussed this part yet.

**Comment by certik** [comment]
I see, nice. I didn't know you can exit from a "block".

What is nice about this approach to exception handling is that there is no extra features needed in the compiler "backend", no long jumps, no stack unwinding, no magic. All this is is just a nice syntax, checked by the compiler's "frontend", but underneath using current Fortran features, so regarding questions like performance there is no additional overhead over what you would need to write by hand anyway.

**Comment by aradi** [comment]
As for the status-proposal: I think, our current approach is superior. There is no scenario, where a program would be error stopped due to an exception. Rather, we would force the programmer to either handle the exception or propagate it upwards. Code, which does not do any of the two (e.g. which tries to ignore an exception) would be simply invalid.

If the programmer wanted to stop the code due to an exception, it would have to happen explicitly, e.g.
```
try catch
  call subroutine_throwing_error(...)
catch all
  error stop "Routine failed and I don't want to deal with it"
end try
```

**Comment by certik** [comment]
Yes, I like this overall design so far. That's where a compiler comes in, to enforce that the error is handled one way or another. The way it is enforced in your library currently is that it will fail at runtime when the error object gets deallocated, which is the only way I think it can be done currently, but if it is part of the language, then it will get enforced at compile time, which I think is much better.

**Comment by aradi** [comment]
@certik In case, you want to prototype it in LFortran, I've found a [blog entry about the error handling implementation in Swift](https://www.mikeash.com/pyblog/friday-qa-2017-08-25-swift-error-handling-implementation.html). I think, this lightweight solution should be exactly, what we should be heading for. 

Actually, thinking about Swifts strategy, we could also consider to use a simple `throws` clause without an explicit list of the error types thrown by the routine:
```fortran
subroutine subroutine_with_error(...) throws
  ...
  throw io_error(name="somefile.dat")
 ...
end subroutine subroutine_with_error
```
This would then simplify interface declarations as well, as one would not have to list all the errors explicitly. And we should go with the Swift convention, that an routine, which does not throw an error, matches an interface, which allows for error throwing (but not the other way around). We should then probably also require, that every `try catch` structure must have a `catch all` clause, unless the embedding routine itself has the `throws` clause. That way, it remains a simple compile time decision, whether a given routine may throw an error or not.

Last, but not least, we could also easily mimic Switfs `try?` and `try!` constructs. By allowing for a  `try ... else ...` in an assignment, we could provide an even superior alternative for their `try?`, where the default value is set explicitely:
```fortran
! some_value contains either the return value of function_throwing_error (if no error occured) or -1 (if error occured)
some_value = try function_throwing_error() else -1
```
And the `try!` construct (which would stop in case the call throws an error), could be probably implemented with a special keyword
```fortran
some_value = try_stop function_throwing_error()
```
The `try_stop` construct could also be used for subroutine calls, if the error should lead to an immediate stop:
```fortran
! Would stop the code, if the subroutine has thrown an error
try_stop call subroutine_throwing_error()
```

**Comment by certik** [comment]
I think the Swift's approach is essentially like Rusts or Zigs, except that Swift handles this automatically, so the error is not exposes explicitly as a type, but the compiler does the equivalent thing: it stores the error as a return type (in a special register) and checks it in the caller. 

That is a design that we can also pursue -- just add some keywords like `throws`, which would instruct the compiler to return the error flag (either success or with an error struct).
